contracts/contracts/GrantRoundManagerUniV2.sol

```diff
@@ -0,0 +1,261 @@
+ * and therefore swaps during donations must be done on a Uniswap V2 fork. Because we are less concerned about
+ * gas costs on these networks, and because there are many different forks that may have slight differences, we
+ * call to an external router instead of inheriting from a router.
+ * @dev Current implementation is hardcoded for Polygon mainnet + SushiSwap
+ */
+contract GrantRoundManagerUniV2 {
+  // --- Libraries ---
+  using Address for address;
+  using SafeERC20 for IERC20;
+  using SafeMath for uint256;
+
+  // --- Data ---
+  /// @notice Address of a router conforming to the Uniswap V2 interface
+  ISwapRouter public constant router = ISwapRouter(0x1b02dA8Cb0d097eB8D57A175b88c7D8b47997506);
+
+  /// @notice WETH address
+  IERC20 public constant WETH = IERC20(0x0d500B1d8E8eF31E21C99d1Db9A6444d3ADf1270);
+
+  /// @notice Address of the GrantRegistry
+  GrantRegistry public immutable registry;
+
+  /// @notice Address of the ERC20 token in which donations are made
+  IERC20 public immutable donationToken;
+
+  /// @dev Used for saving off swap output amounts for verifying input parameters
+  mapping(IERC20 => uint256) internal swapOutputs;
+
+  /// @dev Used for saving off contribution ratios for verifying input parameters
+  mapping(IERC20 => uint256) internal donationRatios;
+
+  /// @dev Scale factor on percentages when constructing `Donation` objects. One WAD represents 100%
+  uint256 internal constant WAD = 1e18;
+
+  /// --- Types ---
+  /// @dev Defines the total `amountIn` of the first token in `path` that needs to be swapped to `donationToken`
+  struct SwapSummary {
+    uint256 amountIn;
+    uint256 amountOutMin; // minimum amount to be returned after swap
+    address[] path; // Use `path == [donationToken]` to indicate no swap is required and just transfer the tokens directly
+  }
+
+  /// @dev Donation inputs and Uniswap V3 swap inputs: https://docs.uniswap.org/protocol/guides/swaps/multihop-swaps
+  struct Donation {
+    uint96 grantId; // grant ID to which donation is being made
+    IERC20 token; // address of the token to donate
+    uint256 ratio; // ratio of `token` to donate, specified as numerator where WAD = 1e18 = 100%
+    GrantRound[] rounds; // rounds against which the donation should be counted
+  }
+
+  // --- Events ---
+  /// @notice Emitted when a new GrantRound contract is created
+  event GrantRoundCreated(address grantRound);
+
+  /// @notice Emitted when a donation has been made
+  event GrantDonation(uint96 indexed grantId, IERC20 indexed tokenIn, uint256 donationAmount, GrantRound[] rounds);
+
+  // --- Constructor ---
+  constructor(GrantRegistry _registry, IERC20 _donationToken) {
+    // Validation
+    require(_registry.grantCount() >= 0, "GrantRoundManager: Invalid registry");
+    require(_donationToken.totalSupply() > 0, "GrantRoundManager: Invalid token");
+
+    // Set state
+    registry = _registry;
+    donationToken = _donationToken;
+
+    // Token approvals of common tokens
```

Comment:
Still testing things out and working on a full review, but quick question: why not just extend from `UniswapV2Router02`, the same way we did for v3, to avoid these kinds of issues? Is it a significant lift technically? Or is there some other downside to doing so?

Reply:
Mainly because UniswapV2 was written with Solidity 0.6.6 😕

Reply:
Got it. I'd be curious what the upgrade would take. But we can hold off for now.

In the meantime, I wonder if instead of pre-approving select tokens in the constructor, we check for existing approval and conditionally approve inside the donate method? This adds some gas (especially for the first donator of a given token) but given we're hardcoding token addresses for Polygon as it is, and Polygon is a very low fee environment, it seems a reasonable tradeoff to make. It also gives us more flexibility token choice wise.

Reply:
> Got it. I'd be curious what the upgrade would take. But we can hold off for now.

Yea I'm also curious, don't recall enough of the breaking changes between the two versions offhand so would have to investigate, probably not worth the time investigating right now until we have a more clear long-term path forward, since like you said the additional cost doesn't matter much on Polygon.

Added in 94446ea542315410bae41811bcae4dc4e8b96bc7